src/collector/mod.rs

```diff
@@ -60,6 +61,70 @@ impl From<CollectorOptions> for WebSocketListenerConfig {
 	}
 }
 
+/// This type is used to distinguish different keys in the storage
+#[derive(Clone, Copy, Debug, Hash, Ord, PartialOrd, Eq, PartialEq)]
+pub(crate) enum CollectorPrefixType {
+	/// Candidate prefixed by ParachainId
+	Candidate(Option<u32>),
+	/// Relay chain block
+	Head,
+}
+
+/// Used to query storage for multiple hash types
+#[derive(Clone, Debug, Ord, PartialOrd, Eq)]
+pub(crate) struct CollectorKey {
+	/// Prefix for a specific collector key
+	pub prefix: CollectorPrefixType,
+	/// Block or a candidate hash
+	pub hash: H256,
+}
+
+impl CollectorKey {
+	pub(crate) fn new_relay_chain_block(hash: H256) -> Self {
+		Self { prefix: CollectorPrefixType::Head, hash }
+	}
+	pub(crate) fn new_parachain_candidate(hash: H256, para_id: u32) -> Self {
+		Self { prefix: CollectorPrefixType::Candidate(Some(para_id)), hash }
+	}
+	// This is used when we don't know the parachain id
+	pub(crate) fn new_generic_hash(hash: H256) -> Self {
+		Self { prefix: CollectorPrefixType::Candidate(None), hash }
+	}
+
+	pub(crate) fn new_with_prefix(prefix: CollectorPrefixType) -> Self {
+		Self { prefix, hash: H256::zero() }
+	}
+}
+
+impl HasPrefix for CollectorKey {
+	fn has_prefix(&self, other: &Self) -> bool {
+		match self.prefix {
+			CollectorPrefixType::Head => other.prefix == self.prefix,
+			CollectorPrefixType::Candidate(maybe_para) => match other.prefix {
+				CollectorPrefixType::Head => false,
+				CollectorPrefixType::Candidate(maybe_other_para) => {
+					maybe_other_para.is_none() || maybe_para.is_none() || maybe_para == maybe_other_para
+				},
+			},
+		}
+	}
+}
+
+// This implementation compares the hashes and optionally checks for the prefix
+// For example, when we have no parachain id available, we should be able to find
+// a specific entry with no hash
+impl PartialEq for CollectorKey {
+	fn eq(&self, other: &Self) -> bool {
+		self.hash == other.hash && (self.has_prefix(other) || other.has_prefix(self))
```

Comment:
Isn't `self.has_prefix(other)` equal to `other.has_prefix(self)` ?

Reply:
It is now, but it wasn't when I've experimented with this code. As previously, that might be better to move this to a separate PR.